apps/term-apply/pkg/applicant/dynamodb.go

```diff
@@ -0,0 +1,293 @@
+package applicant
+
+import (
+	"fmt"
+	"log"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/dynamodb"
+)
+
+/*
+
+The relevant expected schema for the DynamoDB table is as follows:
+
+> applied_date: unix time - number - Sort DDB Key <br>
+> email: candy@date.com - string - Primary/Partition DDB Key <br>
+> github: candydate100 - string - Secondary Global Index <br>
+> name: Candy Date - string <br>
+> role_applied: sr. software engineer - string <br>
+> offer_given: bool <br>
+> rejected: bool <br>
+
+term-apply users have the ability to modify their email after submitting
+an application. If this happens, the existing record will be deleted and
+recreated with updated fields using a transaction to ensure atomic changes.
+Other item edits are done in place.
+
+*/
+
+type emptyResultError struct {
+	user string
+}
+
+func newEmptyResult(user string) *emptyResultError {
+	var err emptyResultError
+	err.user = user
+	return &err
+}
+
+func (err *emptyResultError) Error() string {
+	return fmt.Sprintf("No uploads found for user %s", err.user)
+}
+
+func applicationFromItem(item map[string]*dynamodb.AttributeValue) application {
+	var app application
+
+	appliedDate, exists := item["applied_date"]
+	if exists {
+		app.appliedDate = *appliedDate.N
+	}
+	github, exists := item["github"]
+	if exists {
+		app.github = *github.S
+	}
+	email, exists := item["email"]
+	if exists {
+		app.email = *email.S
+	}
+	name, exists := item["name"]
+	if exists {
+		app.name = *name.S
+	}
+	roleApplied, exists := item["role_applied"]
+	if exists {
+		app.roleApplied = *roleApplied.S
+	}
+	offferGiven, exists := item["offer_given"]
+	if exists {
+		app.offerGiven = *offferGiven.BOOL
+	}
+	rejected, exists := item["rejected"]
+	if exists {
+		app.rejected = *rejected.BOOL
+	}
+
+	return app
+}
+
+// Returns the provided user's most recent application
+func GetApplication(user, table, index string) (application, error) {
+	sess, err := session.NewSessionWithOptions(session.Options{
+		SharedConfigState: session.SharedConfigEnable,
+	})
+	if err != nil {
+		return application{}, err
+	}
+
+	svc := dynamodb.New(sess)
+
+	result, err := svc.Query(&dynamodb.QueryInput{
+		TableName:              aws.String(table),
+		IndexName:              aws.String(index),
+		KeyConditionExpression: aws.String("github = :github"),
+		ExpressionAttributeValues: map[string]*dynamodb.AttributeValue{
+			":github": {S: aws.String(user)},
+		},
+	})
+	if err != nil {
+		return application{}, err
+	}
+
+	if len(result.Items) > 0 {
+		app := applicationFromItem(result.Items[len(result.Items)-1])
+		return app, nil
+	} else {
+		log.Printf("No applications found for %s", user)
+		return application{}, newEmptyResult(user)
+	}
```

Comment:
this else is unnecessary, typically in golang when you return from the first conditional then you can just do what's in the else case as the base case. gopher style is:

```
if something exceptional {
    bail early and return
}
continue doing "happy path case"
```

`else` is usually frowned upon

Reply:
I'll take a look through and see if I can get rid of any else statements :)

Reply:
you can without any changes, since the if condition already returns

Reply:
There's a few else statements still in the project, but they aren't nested (except in the UI code) and removing them might be more verbose. Any egregious or nested else's are gone, reduced to atoms